File: experiments/source/ocaml/clipped_huber_sum.ml
open Loom

let[@loom.entry] clipped_huber_sum (delta : float) (cap : float) (x : Tensor1.t) (y : Tensor1.t) =
  Tensor1.reduce_sum
    (Tensor1.map2
       (fun xi yi ->
         let d = xi -. yi in
         let abs_d = if d >= 0.0 then d else -.d in
         let quadratic = 0.5 *. d *. d in
         let linear = delta *. (abs_d -. (0.5 *. delta)) in
         let value = if abs_d > delta then linear else quadratic in
         if value > cap then cap else value)
       x y)

File: experiments/source/python/clipped_huber_sum.py
import loom


@loom.entry
def clipped_huber_sum(delta: float, cap: float, x: loom.Tensor1, y: loom.Tensor1) -> float:
    def contribution(xi: float, yi: float) -> float:
        d = xi - yi
        abs_d = d if d >= 0.0 else -d
        quadratic = 0.5 * d * d
        linear = delta * (abs_d - (0.5 * delta))
        value = linear if abs_d > delta else quadratic
        return cap if value > cap else value

    return loom.Tensor1.reduce_sum(loom.Tensor1.map2(contribution, x, y))
